source/memtrace.c

```diff
@@ -0,0 +1,488 @@
+        AWS_LOGF_TRACE(AWS_LS_COMMON_MEMTRACE, "Stacks by number of leaks:\n");
+        AWS_LOGF_TRACE(
+            AWS_LS_COMMON_MEMTRACE,
+            "################################################################################\n");
+        aws_hash_table_foreach(&stacks, s_insert_stacks, &stacks_by_count);
+        while (aws_priority_queue_size(&stacks_by_count) > 0) {
+            struct stack_info_t *stack = NULL;
+            aws_priority_queue_pop(&stacks_by_count, &stack);
+            AWS_LOGF_TRACE(AWS_LS_COMMON_MEMTRACE, "%zu allocations leaking %zu bytes:\n", stack->count, stack->size);
+            AWS_LOGF_TRACE(AWS_LS_COMMON_MEMTRACE, "%s\n", (const char *)aws_string_bytes(stack->trace));
+        }
+        aws_priority_queue_clean_up(&stacks_by_count);
+        aws_hash_table_clean_up(&stacks);
+    }
+
+    AWS_LOGF_TRACE(
+        AWS_LS_COMMON_MEMTRACE, "################################################################################\n");
+    AWS_LOGF_TRACE(
+        AWS_LS_COMMON_MEMTRACE, "#  END MEMTRACE DUMP                                                           #\n");
+    AWS_LOGF_TRACE(
+        AWS_LS_COMMON_MEMTRACE, "################################################################################\n");
+
+    aws_mutex_unlock(&tracer->mutex);
+}
+
+static void *s_trace_mem_acquire(struct aws_allocator *allocator, size_t size) {
+    struct alloc_tracer *tracer = allocator->impl;
+    void *ptr = aws_mem_acquire(tracer->allocator, size);
+    s_alloc_tracer_track(tracer, ptr, size);
+    return ptr;
+}
+
+static void s_trace_mem_release(struct aws_allocator *allocator, void *ptr) {
+    struct alloc_tracer *tracer = allocator->impl;
+    s_alloc_tracer_untrack(tracer, ptr);
+    aws_mem_release(tracer->allocator, ptr);
+}
+
+static void *s_trace_mem_realloc(struct aws_allocator *allocator, void *ptr, size_t old_size, size_t new_size) {
+    struct alloc_tracer *tracer = allocator->impl;
+    void *new_ptr = ptr;
+
+    AWS_FATAL_ASSERT(AWS_OP_SUCCESS == aws_mem_realloc(tracer->allocator, &new_ptr, old_size, new_size));
+
+    s_alloc_tracer_untrack(tracer, ptr);
+    s_alloc_tracer_track(tracer, new_ptr, new_size);
+
+    return new_ptr;
+}
+
+static void *s_trace_mem_calloc(struct aws_allocator *allocator, size_t num, size_t size) {
+    struct alloc_tracer *tracer = allocator->impl;
+    void *ptr = aws_mem_calloc(tracer->allocator, num, size);
+    s_alloc_tracer_track(tracer, ptr, num * size);
+    return ptr;
+}
+
+struct aws_allocator *aws_mem_tracer_new(
+    struct aws_allocator *allocator,
+    enum aws_mem_trace_level level,
+    size_t frames_per_stack) {
+    struct aws_allocator *trace_allocator = aws_mem_calloc(aws_default_allocator(), 1, sizeof(struct aws_allocator));
+    AWS_FATAL_ASSERT(trace_allocator);
+    struct alloc_tracer *tracer = aws_mem_calloc(aws_default_allocator(), 1, sizeof(struct alloc_tracer));
```

Comment:
`allocator`

And for bonus points, use acuire_many for the allocator and it's impl.